java/server/src/org/openqa/selenium/grid/sessionqueue/NewSessionQueue.java

```diff
@@ -66,6 +76,40 @@ public boolean hasRequestTimedOut(HttpRequest request) {
     return duration.compareTo(requestTimeout) > 0;
   }
 
+  protected Capabilities validateCaps(Capabilities caps) {
+    String browser = caps.getBrowserName().isEmpty() ? "ANY" : caps.getBrowserName();
+    Platform platform = caps.getPlatformName() == null ? Platform.ANY : caps.getPlatformName();
+    String version = caps.getBrowserVersion().isEmpty() ? "ANY" : caps.getBrowserVersion();
+
+    return new ImmutableCapabilities(
+      CapabilityType.BROWSER_NAME, browser,
+      CapabilityType.PLATFORM_NAME, platform,
+      CapabilityType.BROWSER_VERSION, version);
+  }
+
+  protected void setCount(Capabilities caps, Map<String, BrowserInfo> browserInfoMap) {
```

Comment:
Making things immutable will require you to rework this a little. My apologies for the extra work.